pmlb/pmlb.py

```diff
@@ -137,3 +137,89 @@ def get_updated_datasets(local_cache_dir='datasets'):
     return {'changed_datasets': changed_datasets,
             'changed_metadatas': changed_metadatas}
 
+from sklearn.neighbors import NearestNeighbors
+from sklearn.preprocessing import StandardScaler
+from .update_dataset_files import generate_summarystats
+from pathlib import Path
+import numpy as np
+
+
+def fetch_nearest_dataset_names(X,y=None, **kwargs):
+    """
+    X: numpy array
+        an n_samples x n_features array of independent variables
+    y: numpy array or None (default: None)
+        a n_samples array of dependent variables
+    """
+    df = pd.DataFrame({**{'x_'+str(i):x for i,x in enumerate(X.transpose)}
+                       **{'target':y}})
+    return fetch_nearest_dataset_names(df, **kwargs)
+
+def fetch_nearest_dataset_names(df, n=1, 
+        dimensions=['#instances','#features'],
+        task=None):
+    """Returns names of most similar datasets to df, in order of similarity. 
+
+    Parameters
+    ----------
+    df: pandas Dataframe 
+        a dataframe of n_samples x n_features+1 with a target column labeled
+        'target'
+    n: int (default: 1)
+        the number of dataset names to return
+    dimensions: list of str or str (default: ['NumberOfInstances',
+    'NumberOfFeatures'])
+        a list of dataset characteristics to include in similarity calculation.
+        Dimensions must correspond to columns of datasets/all_summary_stats.csv.
+        If 'all', uses all numeric columns.
+    task: str or None (default: None)
+        specify classification or regression for summary stat generation. If 
+        None, we use classification unless the target column has more than 5 
+        unique values.
+
+    Returns
+    -------
+    dataset_names: an n-element list of dataset names in order of most similar 
+        to least similar.
+    """
+    # get summary stats for dataset
+    if task==None:
+        task = 'classification' if df['target'].nunique()<5 else 'regression'
+
+    # load pmlb summary stats
+    path = Path(__file__).parent / "all_summary_stats.tsv"
+    pmlb_stats = pd.read_csv(path)
+    all_names = pmlb_stats['dataset'].values
+    # restrict to same task
+    pmlb_stats = pmlb_stats.loc[pmlb_stats.problem_type==task]
+    # restrict to floating point data in stats
+    pmlb_stats = pmlb_stats.apply(
+            lambda x: pd.to_numeric(x,errors='coerce')).dropna(axis=1,how='all')
+
+
+    if dimensions=='all':
+        dimensions = list(pmlb_stats.columns)
+    else:
+        pmlb_stats = pmlb_stats[dimensions]  
+        assert(all([d in pmlb_stats.columns for d in dimensions]))
+
+    dataset_stats = generate_summarystats(df, 'dataset', task)
```

Comment:
I'm not sure if we want to generate new summary statistics in this function.

Reply:
Oh I see now. `df` does not have to be one of PMLB datasets. Got it!

Reply:
this function was changed in my PR to return the dataframe

Reply:
I must have missed it, but we would also need to import [`get_dataset_stats`](https://github.com/EpistasisLab/penn-ml-benchmarks/blob/d7b11072a9beb2740c5bce26c4ae8e391525ebb9/pmlb/update_dataset_files.py#L106) for this to work.

Reply:
These updates were pushed when I added the autogeneration to CI. Didn't mean to make things hard for you. My bad!

Reply:
[This tweak](https://github.com/EpistasisLab/penn-ml-benchmarks/commit/03831e723bd61ae46e781f7111a810899aa002cc) returns the data frame in the function.